filecoin-proofs/src/api/mod.rs

```diff
@@ -449,140 +77,78 @@ pub fn get_unsealed_range<T: Into<PathBuf> + AsRef<Path>>(
-    {
-        let (_, mut aligned_source) = get_aligned_source(&mut source, &piece_lengths, piece_size);
-        write_padded(&mut aligned_source, target)
-    }
-
-    #[test]
-    fn test_generate_piece_commitment_up_to_minimum() -> Result<(), failure::Error> {
-        // test comm_p generation for all byte lengths up to the minimum piece alignment when
-        // writing a piece to a sector
-        let max = MINIMUM_PIECE_SIZE as usize;
-
-        for n in 0..=max {
-            let bytes: Vec<u8> = (0..n).map(|_| rand::random::<u8>()).collect();
-            let mut data_a = vec![0; n];
-            let mut data_b = vec![0; max];
-
-            for i in 0..n {
-                data_a[i] = bytes[i];
-                data_b[i] = bytes[i];
-            }
-
-            let comm_p_a = generate_comm_p(&data_a)?;
-            let comm_p_b = generate_comm_p(&data_b)?;
-
-            assert_eq!(comm_p_a, comm_p_b);
-        }
+// Takes a piece and the size of the piece and returns the comm_p.
+pub fn generate_piece_commitment<T: std::io::Read>(
+    unpadded_piece_file: T,
+    unpadded_piece_size: UnpaddedBytesAmount,
+) -> error::Result<PieceInfo> {
+    let mut padded_piece_file = tempfile()?;
+    let unpadded_piece_size_with_alignment = add_piece(
+        unpadded_piece_file,
+        &mut padded_piece_file,
+        unpadded_piece_size,
+        &[],
+    )?;
 
-        Ok(())
-    }
+    let _ = padded_piece_file.seek(SeekFrom::Start(0))?;
 
-    #[test]
-    fn test_generate_piece_commitment_over_minimum() -> Result<(), failure::Error> {
-        // sanity check that larger byte lengths are still zero padded
-        let bytes: Vec<u8> = (0..400).map(|_| rand::random::<u8>()).collect();
-        let mut data_a = vec![0; 400];
-        let mut data_b = vec![0; 508];
-
-        for i in 0..400 {
-            data_a[i] = bytes[i];
-            data_b[i] = bytes[i];
-        }
+    let commitment =
+        generate_piece_commitment_bytes_from_source::<DefaultPieceHasher>(&mut padded_piece_file)?;
 
-        let comm_p_a = generate_comm_p(&data_a)?;
-        let comm_p_b = generate_comm_p(&data_b)?;
+    Ok(PieceInfo {
+        commitment,
+        size: unpadded_piece_size_with_alignment,
+    })
+}
 
-        assert_eq!(comm_p_a, comm_p_b);
+/// Write a piece. Returns the unpadded, but aligned size of the piece.
+pub fn add_piece<R, W>(
+    source: R,
+    target: W,
+    piece_size: UnpaddedBytesAmount,
+    piece_lengths: &[UnpaddedBytesAmount],
+) -> error::Result<UnpaddedBytesAmount>
+where
+    R: Read,
+    W: Read + Write + Seek,
+{
+    ensure!(
+        piece_size >= UnpaddedBytesAmount(MINIMUM_PIECE_SIZE),
+        "Piece must be at least {} bytes",
+        MINIMUM_PIECE_SIZE
+    );
+    let padded_piece_size: PaddedBytesAmount = piece_size.into();
+    ensure!(
+        u64::from(padded_piece_size).is_power_of_two(),
+        "Aligned and padded piece size must be a power of 2 ({:?})",
```

Comment:
BLOCKING: This error message doesn't make much sense to me, and tripped me up when I was integrating with the sector builder. 

The caller doesn't control adding alignment bytes (this function adds them).

Perhaps this error message should be:

```
Bit-padded piece size must be a power of 2 ({:?})
```

Reply:
fixed